src/vc.rs

```diff
@@ -304,6 +310,9 @@ pub struct LinkedDataProofOptions {
     #[serde(skip_serializing_if = "Option::is_none")]
     /// Checks to perform
     pub checks: Option<Vec<Check>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    /// Additional proof data parameters
+    pub property_set: Option<Map<String, Value>>,
```

Comment:
`LinkedDataProofOptions` was supposed to be compatible with `vc-http-api`, which used JSON Schema with `additionalProperties: false`. I think there is a `vc-http-api` test that will fail if the implementation allows unknown properties. This could be addressed by adding the needed properties to the struct specifically, i.e. `capability` and/or `capability_chain`.

`vc-http-api` since split `LinkedDataProofOptions` into multiple types, e.g. [IssueCredentialOptions](https://github.com/w3c-ccg/vc-http-api/blob/50b9d64/docs/components/IssueCredentialOptions.yml) and [PresentCredentialOptions](https://github.com/w3c-ccg/vc-http-api/blob/50b9d64/docs/components/PresentCredentialOptions.yml), but those are still using `additionalProperties: false`.

Reply:
what do you think about making `Proof` and `LinkedDataProofOptions` generic, such that the `vc-http-api` proof options can be something like `type VCHtttpApiOptions = LinkedDataProofOptions<()>;`. I experimented with this in the `feat/generic_proof` branch (until the refactor got out of control), essentially because VC proofs and ZCAP proofs are both subsets of LD proofs.

Reply:
I had a look but I'm not quite sure of the benefit (except maybe that the types relations are clearer). Is it mainly to get rid of the map for "extra properties"? But it's still a map so I guess not. Also the proof purpose is still an enum and I don't think you can make a generic over an enum variant? I know this is for another time, just writing my questions down before I forget them when we have that conversation.

Reply:
I like the idea of making LD-Proofs generic over both VCs and ZCAPs--sounds like a good architecture conformant to specs. Using generics to slot in extra fields could be overkill, but I'll let people here decide.

Reply:
Making things more generic sounds good; it would be nice to reduce some of the code and logic duplication around proof generation and verification. I'm not sure I fully understand the idea though.

I understand the Proof object and proof options to be determined by the Proof type more than by the outer document type. The `capability` property looks like an exception - are there other document-type-specific proof properties? If not, could `capability` be added to the Proof struct directly? On the other hand, maybe it would be useful to express this relationship in the types, just in case.

Proof Purpose seems like a "core" thing to linked data proofs so I'm not sure I understand making it generic. However, some document types more or less use specific proof purposes. i.e. VC uses assertionMethod and VP uses authentication - by default and/or convention, but I understand they could use other proof purposes also. Invocation and Capability documents, however, must use capabilityInvocation and capabilityDelegation proof purposes, respectively. Could this be expressed via associated types in LinkedDataDocument for default and required proof purpose?

Maybe the proof object could be connected to the ProofSuite type somehow? e.g. the `eip712Domain` property is only expected for the `EthereumEip712Signature2021` proof. Some of our proof types use publicKeyJwk in the proof...

> `type VCHtttpApiOptions = LinkedDataProofOptions<()>;`

Newer VC HTTP API has IssueCredentialOptions, PresentCredentialOptions, and VerifyOptions; instead of LinkedDataProofOptions. Could it be useful to work with these newer more differentiated types instead? Although we might find something like LinkedDataProofOptions reappear in other linked data signatures work.

I wonder how using a generic `property_set` would compare with composing structs. e.g.:
```rust
pub struct EthereumEip712Signature2021 {
  #[serde(flatten)]
  pub base: Proof,
  pub eip712_domain: Option<Eip712ProofDomain>,
}
impl Verify for EthereumEip712Signature {
  // ...
}
```
Unfortunately that conflicts with the current type definitions.

Reply:
After thinking about it more, I'm okay with this change (adding property_set to LinkedDataProofOptions).
The mentioned `vc-http-api` test is no longer present, as noted in https://github.com/w3c-ccg/vc-http-api/pull/142#issuecomment-793960808. We seem to need the API to allow for proof-type-specific proof options (e.g. `eip712Domain`) as well as proof-purpose-specific options (e.g. `capability`; previously I thought this was document-type-specific, but it's simpler to consider it proof-purpose-specific - and capability documents might not have a `type` anyway). However, I don't think `property_set` should be passed through into the proofs. There may be issue options that we don't support or know about but that are supposed to have a meaning other than passing them through into the proof. e.g. `credentialStatus` issue option is supposed to cause a `credentialStatus` object to be added to the document rather than the proof: https://github.com/w3c-ccg/vc-http-api/blob/7c72ae2b470ff8dfec12d3b6c3de4f84c6f2a176/packages/vc-http-api-test-server/__tests__/issueCredential.spec.js#L130-L148